library/std/src/thread/mod.rs

```diff
@@ -718,7 +723,32 @@ pub(crate) fn set_current(thread: Thread) {
 /// In contrast to the public `current` function, this will not panic if called
 /// from inside a TLS destructor.
 pub(crate) fn try_current() -> Option<Thread> {
-    CURRENT.try_with(|current| current.get_or_init(|| Thread::new_unnamed()).clone()).ok()
+    CURRENT
+        .try_with(|current| {
+            current
+                .get_or_init(|| {
+                    let thread = Thread::new_unnamed();
+                    CURRENT_ID.set(Some(thread.id()));
+                    thread
+                })
+                .clone()
+        })
+        .ok()
+}
+
+/// Gets the id of the thread that invokes it.
+///
+/// If called from inside a TLS destructor and the thread was never
+/// assigned an id, returns `None`.
+#[inline]
+pub(crate) fn try_current_id() -> Option<ThreadId> {
```

Comment:
That's not true. If `CURRENT` hasn't been initialized yet, it can be initialized, even in a TLS destructor. If it has been initialized, and we are inside its destructor, then `CURRENT_ID` has been set already. So the only place where this can fail is if the global allocator calls this function while trying to deallocate the TLS storage for the current ID. The global allocator shouldn't be doing that anyway, so we can just panic in that case. Therefore, this function can be infallible (I slightly prefer the name `id`, like with `current` the module and signature alone suffice IMHO):
```suggestion
#[inline]
pub(crate) fn id() -> ThreadId {
```